### PR TITLE
Add optional alpha-threshold passthrough to the UI picking backend

### LINKED ISSUE #17448: Add optional alpha / transparency passthrough for UI picking backend
## What problem does this solve or what need does it fill?

Optional transparency passthrough was added to the sprite picking backend in #16388.

## What solution would you like?

- I'd like to have the same capability for the UI picking backend. I have some decorative UI panels and I'd like avoid picking triggering when hovering their transparent areas.
- This was also hinted at in https://github.com/bevyengine/bevy/issues/14941#issuecomment-2313653684 but perhaps we forgot to add a tracking issue.
- As a bonus: Pending blessing, I'd consider the implementor doing a per-entity solution from day 1, so that you can mark some entities as using the alpha threshold, while others just inherit the global resource. If I had to guess, it doesn't seem much work adding that at the same time, but time will tell.
- Implementation to be verified but maybe spawning an entity with per-entity picking settings would look like this:

```rust
commands
.spawn((
    Node {
        // ...
        ..Default::default()
    },
    ImageNode {
        // ...
        ..Default::default()
    },
    ImagePickingSettings { picking_mode: ImagePickingMode::BoundingBox },
));
```

- Noting that, for the sprite picking backend, `SpritePickingSettings` defaults to `{ picking_mode: SpritePickingMode::AlphaThreshold(0.1) }` and so the example here is to illustrate that this entity is _overriding_ the OOTB alpha-threshold check to instead use a bounding-box check.

## What alternative(s) have you considered?

If you just have rounded buttons maybe #14957 already works for you, but it doesn't in my case.

### PR DESCRIPTION
**Pull request by mgi388**
# Objective

- Fixes #17448, see that for more info.
- tl;dr: Avoid picking triggering when hovering transparent areas of UI images, like decorative panels.

## Solution

- Port the work done by @vandie in #17448 to `bevy_ui`.

## Testing

- Added a new example, similar to the `sprite_picking` example:

```
cargo run --example ui_image_picking --features ui_picking
```

---

## Showcase

https://github.com/user-attachments/assets/3ab4ea87-67c2-4ea5-929b-e54ae6b5a7b4

### CODE REVIEW

**Comment by vandie** [comment]
Probably should have done this back when we did the non-ui version. Sorry that's on me for not thinking 😂 Nice one!

**Review by vandie — APPROVED** [review]
Nice Job!

**Comment by mgi388** [comment]
> Probably should have done this back when we did the non-ui version. Sorry that's on me for not thinking 😂 Nice one!

Not at all! You deserve the credit here and paved the way for this so thank you!

File: _release-content/migration-guides/ui_picking_alpha_threshold.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+---
+title: UI picking now ignores transparent pixels of image nodes
+pull_requests: [25077]
+---
+
+UI picking now ignores transparent regions of `ImageNode`s by default, matching the behavior of
```
Comment (alice-i-cecile) [inline _release-content/migration-guides/ui_picking_alpha_threshold.md:6]: Good choice of default!

File: _release-content/migration-guides/ui_picking_alpha_threshold.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+sprite picking. A pointer over a pixel of an image node with an alpha value less than or equal to
+`0.1` will no longer register that node as a hit.
+
+This is controlled by the new `UiPickingSettings::picking_mode` field, which defaults to
+`UiPickingMode::AlphaThreshold(0.1)`.
+
+To restore the previous behavior (picking based on the node's bounding box, regardless of
```
Comment (alice-i-cecile) [inline _release-content/migration-guides/ui_picking_alpha_threshold.md]:
```suggestion
To restore the (slightly faster) previous behavior (picking based on the node's bounding box, regardless of
```

File: crates/bevy_ui/src/picking_backend.rs [hunk 3]
```diff
@@ -41,6 +44,29 @@ use bevy_picking::backend::prelude::*;
 #[reflect(Debug, Default, Component)]
 pub struct UiPickingCamera;
 
+/// An optional component that specifies how [`UiPickingPlugin`] should handle
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:47]: This could use a breadcrumb in the other direction too :)

File: crates/bevy_ui/src/picking_backend.rs [hunk 4]
```diff
@@ -41,6 +44,29 @@ use bevy_picking::backend::prelude::*;
...
+/// picking for a UI node, in particular how it treats the transparent pixels of
+/// an [`ImageNode`].
+///
+/// The picking mode can be set globally via
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:51]: These docs should call out what the default behavior is explicitly.

File: crates/bevy_ui/src/picking_backend.rs [hunk 5]
```diff
@@ -229,16 +262,42 @@ pub fn ui_picking(
...
-                            node.pickable.cloned(),
-                            node.transform.inverse().transform_point2(*cursor_position)
-                                / node.node.size(),
-                        ));
+                    // Normalized cursor position relative to the node, with
+                    // `(-0.5, -0.5)` at the top left and `(0.5, 0.5)` at the
+                    // bottom right.
+                    let relative_cursor_position =
+                        node.transform.inverse().transform_point2(*cursor_position)
+                            / node.node.size();
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:272]: Is this right with respect to borders and padding? I would like some tests to verify: it's easy for this to get confused.

File: examples/picking/ui_image_picking.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            align_items: AlignItems::Center,
+            row_gap: px(10),
+        }
+        Children [
+            Text(label),
+            (
+                // A backing panel that fills the node's rectangle, so the
+                // transparent parts of the image are visible against it.
+                ImageNode { image: "branding/bevy_bird_dark.png" }
+                BackgroundColor(Color::srgb(0.15, 0.15, 0.15))
```
Comment (alice-i-cecile) [inline examples/picking/ui_image_picking.rs:79]: We should account for the effects of the BackgroundColor as well, right? Can you add a test to make sure that works?

File: crates/bevy_ui/src/picking_backend.rs [hunk 7]
```diff
@@ -279,6 +338,73 @@ pub fn ui_picking(
...
+    }
+    if image_node.flip_y {
+        uv.y = 1.0 - uv.y;
+    }
+
+    let texture_position = texture_rect.min + uv * texture_rect.size();
+
+    let Ok(color) = image.get_color_at(texture_position.x as u32, texture_position.y as u32) else {
+        // We don't know how to interpret the pixel.
+        return false;
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:404]: We should at least `warn!` here.

File: crates/bevy_ui/src/picking_backend.rs [hunk 8]
```diff
@@ -299,3 +425,245 @@ fn pick_ui_text_section(
         .get(section_index as usize)
         .map(|e| e.entity)
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:432]: Please add a test for 1 pixel-wide UI, both opaque and transparent :) That's a reasonably important case, and an easy edge case.

Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:432]: These tests need more coverage of the node -> correct UV path.

File: crates/bevy_ui/src/picking_backend.rs [hunk 9]
```diff
@@ -279,6 +338,73 @@ pub fn ui_picking(
     }
 }
 
+/// Returns `true` if the pixel of `image_node`'s texture under `uv` (normalized
+/// `0..1` with `(0, 0)` at the top left) has an alpha value greater than
+/// `cutoff`.
+///
+/// Nodes whose image can't be sampled are treated as hits. This includes nodes
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/picking_backend.rs:347]: The semantics feel weird here. Should we be returning something other than a `bool`?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
More cleanup to do, and a couple of possible bugs <3 More tests please!

File: crates/bevy_ui/src/picking_backend.rs (hunk 7 above)
Comment (mgi388) [reply crates/bevy_ui/src/picking_backend.rs:404]: I believe this will be too spammy if we log anything here but I'll double check when I resolve the other feedback.

File: _release-content/migration-guides/ui_picking_alpha_threshold.md (hunk 2 above)
Comment (mgi388) [reply _release-content/migration-guides/ui_picking_alpha_threshold.md]: https://github.com/bevyengine/bevy/pull/25077/changes/2232a0458dd31ff057a4d761384ffac84104b3da

File: crates/bevy_ui/src/picking_backend.rs (hunk 3 above)
Comment (mgi388) [reply crates/bevy_ui/src/picking_backend.rs:47]: https://github.com/bevyengine/bevy/pull/25077/changes/8f9a5744d0d7c3fb460178d471f5816faf619ee0

File: crates/bevy_ui/src/picking_backend.rs (hunk 4 above)
Comment (mgi388) [reply crates/bevy_ui/src/picking_backend.rs:51]: https://github.com/bevyengine/bevy/pull/25077/changes/dea8cd16822210164c4ee7e8ea3cacb4286d5839

File: crates/bevy_ui/src/picking_backend.rs (hunk 8 above)
Comment (mgi388) [reply crates/bevy_ui/src/picking_backend.rs:432]: https://github.com/bevyengine/bevy/pull/25077/changes/c627e788e46153bbe9793eafdec48de453d7770c

**Comment by mgi388** [comment]
@alice-i-cecile I'm still working on resolving the rest of your comments, so I'm reverting the label change if that's OK :)
